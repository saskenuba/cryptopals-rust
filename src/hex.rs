use std::borrow::Cow;
use std::fmt::{Debug, Formatter};
use std::ops::Deref;

use bitvec::field::BitField;
use bitvec::order::Msb0;
use bitvec::prelude::AsBits;

use crate::B64;

// A valid hexadecimal string as bytes
pub struct Hex<'a>(Cow<'a, [u8]>);

impl<'a> Deref for Hex<'a> {
    type Target = Cow<'a, [u8]>;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

impl<'a> Debug for Hex<'a> {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(f, "{:#x?}", self.0)
    }
}

pub trait ToHex {
    type Error;

    fn to_hex<'a>(&self) -> Result<Hex<'a>, Self::Error>;
}

impl<T> ToHex for T
where
    T: AsRef<[u8]>,
{
    type Error = &'static str;

    fn to_hex<'a>(&self) -> Result<Hex<'a>, Self::Error> {
        let bytes_iter = self.as_ref().chunks_exact(2);
        let mut buffer = Vec::with_capacity(self.as_ref().len());

        for bytes in bytes_iter {
            let to_str = std::str::from_utf8(bytes).map_err(|_| "invalid ")?;
            let hex = u8::from_str_radix(to_str, 16).map_err(|_| "invalid hex bytes")?;
            buffer.push(hex);
        }
        Ok(Hex(Cow::from(buffer)))
    }
}

impl<'a> Hex<'a> {
    pub fn into_b64_string(self) -> String {
        let mut buffer = String::with_capacity(self.0.len());
        self.0.as_bits::<Msb0>().chunks(6).for_each(|bits| {
            let encoded = B64.chars().nth(dbg!(bits.load_be())).unwrap();
            buffer.push(encoded);
        });
        buffer
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn success_b64_conversion() {
        let source = "49276d206b696c6c696e6720796f757220627261696e206c696b65206120706f69736f6e6f7573206d757368726f6f6d";
        let expected = "SSdtIGtpbGxpbmcgeW91ciBicmFpbiBsaWtlIGEgcG9pc29ub3VzIG11c2hyb29t";
        insta::assert_snapshot!(source.to_hex().unwrap().into_b64_string());
    }
    #[test]
    fn success_parse_str_as_hex_str() {
        let source = "49276d";
        let expected = &[0x49u8, 0x27, 0x6d];

        let res = source.to_hex();
        insta::assert_snapshot!(format!("{:?}", res));
    }
}
