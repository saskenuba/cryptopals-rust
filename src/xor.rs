use std::ops::BitXor;

use anyhow::ensure;
use bitvec::prelude::*;

use crate::AnyResult;

pub fn fixed_xor<T1, T2>(b1: T1, b2: T2) -> AnyResult<Vec<u8>>
where
    T1: AsRef<[u8]>,
    T2: AsRef<[u8]>,
{
    let b1 = b1.as_ref();
    let b2 = b2.as_ref();

    ensure!(b1.len() == b2.len());
    let buffer: BitVec<_> = b1
        .view_bits::<Lsb0>()
        .iter()
        .zip(b2.view_bits::<Lsb0>().iter())
        .map(|(bb1, bb2)| *bb1 ^ *bb2)
        .collect();

    Ok(buffer.into_vec())
}
pub fn fixed_xor_byte<T1>(b1: T1, b2: u8) -> AnyResult<Vec<u8>>
where
    T1: AsRef<[u8]>,
{
    let b1 = b1.as_ref();
    let buffer: BitVec<_> = b1.iter().map(|bb1| bb1.bitxor(b2)).collect();

    Ok(buffer.into_vec())
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::hex::ToHex;

    #[test]
    fn success_xor_two_buffers() {
        let source = "1c0111001f010100061a024b53535009181c";
        let xor_against_hex = "686974207468652062756c6c277320657965"
            .as_bytes()
            .to_hex()
            .unwrap();

        let source_hex_bytes = source.as_bytes().to_hex().unwrap();
        let res = fixed_xor(&source_hex_bytes, &xor_against_hex).unwrap();

        let expected = b"746865206b696420646f6e277420706c6179"
            .as_slice()
            .to_hex()
            .unwrap();
        assert_eq!(*expected, res);
    }

    #[test]
    fn a() {
        let a = 10 ^ 10;
    }
}
