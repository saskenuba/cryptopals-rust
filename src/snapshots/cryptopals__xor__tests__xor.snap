---
source: src/xor.rs
expression: "std::str::from_utf8(&res).unwrap()"
---
the kid don't play
